internal/api/token_refresh.go

```diff
@@ -120,6 +107,60 @@ func (a *API) RefreshTokenGrant(ctx context.Context, w http.ResponseWriter, r *h
 				return internalServerError(terr.Error())
 			}
 
+			if a.config.Sessions.SinglePerUser {
+				sessions, terr := models.FindAllSessionsForUser(tx, user.ID, true /* forUpdate */)
+				if models.IsNotFoundError(terr) {
+					// because forUpdate was set, and the
+					// previous check outside the
+					// transaction found a user and
+					// session, but now we're getting a
+					// IsNotFoundError, this means that the
+					// user is locked and we need to retry
+					// in a few milliseconds
+					retry = true
+					return terr
+				} else if terr != nil {
+					return internalServerError(terr.Error())
+				}
+
+				sessionTag := session.DetermineTag(config.Sessions.Tags)
+
+				// go through all sessions of the user and
+				// check if the current session is the user's
+				// most recently refreshed valid session
+				for _, s := range sessions {
+					if s.ID == session.ID {
+						// current session, skip it
+						continue
+					}
+
+					if s.CheckValidity(retryStart, nil, config.Sessions.Timebox, config.Sessions.InactivityTimeout) != models.SessionValid {
+						// session is not valid so it
+						// can't be regarded as active
+						// on the user
+						continue
+					}
+
+					if s.DetermineTag(config.Sessions.Tags) != sessionTag {
+						// if tags are specified,
+						// ignore sessions with a
+						// mismatching tag
+						continue
+					}
+
+					// since token is not the refresh token
+					// of s, we can't use it's UpdatedAt
+					// time to compare!
+					if s.LastRefreshedAt(nil).After(session.LastRefreshedAt(&token.UpdatedAt)) {
```

Comment:
how do we know tht the token is not the refresh token of s here? is it because of the check above where we compare if `s.ID == sessionID` so we know that at this point, the refresh token used belongs to a different session?

Reply:
I added the comment to justify the `nil` in `LastRefreshedAt()`. The idea is to signal that you shouldn't use `token.UpdatedAt` because `token` does not belong to `s` (one of the _other_ sessions belonging to the user).

Reply:
hmm maybe we can also move this block under an `else` above so something like:
```
if s.ID == sessionID {
 ...
} else {
  if s.LastRefreshedAt(nil).After(session.LastRefreshedAt(&token.UpdatedAt)) {
    ...
  }
}
```